rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,430 @@
+on cache-coherent hardware where native loads and stores have `Relaxed` or
+stronger semantics. The vast majority of hardware which Rust supports today and
+is expected to support in the future is cache-coherent, and therefore making
+volatile feel more at home on such hardware is a major win.
+
+However, we obviously must still have something for those exotic platforms whose
+basic memory loads and stores are not cache-coherent, such as `nvptx`. Hence the
+compromise of `load_not_atomic()` and `store_not_atomic()` is still kept around,
+only discouraging its use.
+
+---
+
+It should be noted that switching to LLVM's atomic volatile accesses does not
+resolve the second problem very well per se, because although oversized volatile
+accesses will not compile anymore, they will only be "reported" to the user via
+LLVM crashes. This is not a nice user experience, which is why volatile wrapper
+types are proposed.
+
+Their design largely mirrors that of Rust's existing atomic types, which is only
+appropriate since they do expose atomic operations. One goal of this design is
+that it should be possible to re-use the architecture-specific code that already
+exists to selectively expose Rust atomic types and operations depending on what
+the hardware supports under the hood.
+
+---
+
+This RFC currently proposes to expose the full power of LLVM's atomic volatile
+operations, including e.g. read-modify-write operations like compare-and-swap,
+because there are legitimately useful use cases for these operations in
+interprocess communication scenarios.
+
+However, the fact that these operations do not necessarily compile into a single
+hardware instruction is arguably a footgun for volatile's use cases, and it
+could be argued that initially only stabilizing loads, stores and `Relaxed`
+atomic ordering would be more prudent. I'll go back to this in the alternatives
+section of this RFC.
+
+---
+
+As currently designed, this RFC uses `arbitrary_safe_types` to give method-like
+semantics to a `NonNull` raw pointer. I believe that this is necessary to get
+reasonable ergonomics with an atomics-like wrapper type approach. However, it
+could also be argued that a `VolatileXyz::store(ptr, data, ordering)` style of
+API could reasonably work, and avoid coupling with unstable features. Similarly,
+the use of `NonNull` itself could be debated. I'll come back to these points in
+the alternatives section as well.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Deprecating the existing volatile operations will cause language churn. Not
+deprecating them will keep around two different and subtly incompatible ways to
+do the same thing, which is equally bad. It could be argued that the issues with
+the existing volatile operations, while real, do not warrant the full complexity
+of this RFC.
+
+Atomic volatile operations are also a somewhat LLVM-specific concept, and
+requiring them may make the life of our other compiler backends harder.
```

Comment:
cc @bjorn3 @sunfish

Reply:
I think that atomic volatile operations would be as easy to support as normal atomics. If I understand it correctly, volatile should only prevent removing loads and stores.

Reply:
It also prevents reordering volatile operations with each other, kind of like `SeqCst` in the atomic world.

Reply:
Actually atomic volatile operations are part of the C++11 standard. If you look at the [operations on std::atomic](https://en.cppreference.com/w/cpp/atomic/atomic/exchange) there are overloads for volatile and non-volatile atomics.

Reply:
> Actually atomic volatile operations are part of the C++11 standard. If you look at the operations on std::atomic there are overloads for volatile and non-volatile atomics.

That's correct, they are also part of the C11 standard: if you look at the atomic operations, e.g., [`atomic_load`](https://en.cppreference.com/w/c/atomic/atomic_load) you'll see that they take a `volatile*` (really).

Reply:
Good point! Let me amend that part...

Reply:
...fixed !

Reply:
Since "atomic volatile" support is actually required for C11/C++11 atomics support, I think my initial assertion that this is an LLVM-specific trick which may make the life of other backends harder was mistaken (if they want to support Rust, they probably want to support modern C/++ as well anyway), and this discussion can therefore be closed.